### PR TITLE
[bugfix] Fix dataloading for iterable datasets and limit_train_batches

### LINKED ISSUE #6332: Validation not called when using an IterableDataset and `limit_train_batches` flag
When using an IterableDataset the `limit_train_batches` trainer flag will prevent the validation loop from being called. Here is  [a simplified notebook to reproduce the issue](https://colab.research.google.com/drive/1TMQigHK4G6egC1IXj-YnRu0P-L3zAA_K?usp=sharing). Note this error may manifest as:

> ValueError: cannot convert float NaN to integer 

if only `limit_train_batches` is set, but can also interfere with model checkpointing if the `limit_val_batches` flag is set: 

> MisconfigurationException: ModelCheckpoint(monitor='val_log') not found in the returned metrics: ['loss']. HINT: Did you call self.log('val_log', tensor) in the LightningModule?

**Possible Workaround:**
Interestingly if the IterableDataset has `__len__` defined than neither flag is an issue.

**Environment Info:**

* PyTorch Version (e.g., 1.0): 1.7.1+cu101
* OS (e.g., Linux): Linux
* How you installed PyTorch (conda, pip, source): pip
* Build command you used (if compiling from source):
* Python version: 3.8.2
* Any other relevant information: 32 cores, 128GB mem
* Pytorch Lightning Version: 1.1.5

### LINKED ISSUE #7303: NAN for progress bar when only `limit_train_batches` is set for trainer with datasets with no __len__
## Please reproduce using the BoringModel

when only `limit_train_batches` is set for trainer for Iterable dataset with no `__len__`, it does not work for progress bar, as the progress bar's total number would be nan, hence https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/progress.py#L412 would crash

### To Reproduce
```
import torch
import pytorch_lightning as pl
from torch.utils.data import Dataset, IterableDataset, DataLoader


def example():
    return {"x": torch.randn(20), "y": torch.ones(1).long()}


class RegularIterableDataset(IterableDataset):
    def __init__(self, num_examples):
        super().__init__()
        self.num_examples = num_examples
        
    def __iter__(self):
        for _ in range(self.num_examples):
            yield example()


import torch.nn.functional as F


class BasicModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(20, 10)

    def forward(self, x):
        return torch.relu(self.l1(x))

    def training_step(self, batch, batch_idx):
        x = batch["x"]
        y = batch["y"]
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y.view(-1))
        self.log("loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        return self.training_step(batch, batch_idx)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.02)

import inspect
from pytorch_lightning.callbacks.progress import ProgressBar
import math


class ProgressBarCheckingCallback(ProgressBar):
    def on_train_start(self, trainer, pl_module):
        super().on_train_start(trainer, pl_module)
        if self.main_progress_bar.total is not None:
            assert not math.isnan(self.main_progress_bar.total)
        
    def on_train_epoch_start(self, trainer, pl_module):
        super().on_train_epoch_start(trainer, pl_module)
        if self.main_progress_bar.total is not None:
            assert not math.isnan(self.main_progress_bar.total) # <- this assertion breaks, which will cause the on_train_batch_end break
        
    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
        super().on_train_batch_end(trainer, pl_module, outputs, batch, batch_idx, dataloader_idx) # <- because of the nan bug: https://fburl.com/j1s8uvum will crash
    
trainer = pl.Trainer(
    max_epochs=2,
    limit_train_batches=3,
    callbacks=[
        ProgressBarCheckingCallback(refresh_rate=1, process_position=0)
    ]
)
model = BasicModel()
train_dl = DataLoader(dataset=RegularIterableDataset(num_examples=128), batch_size=8)
val_dl = DataLoader(dataset=RegularIterableDataset(num_examples=128), batch_size=8)

trainer.fit(model, train_dataloader=train_dl, val_dataloaders=val_dl)


```

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #7303
Fixes #6332 

The check in the training loop for whether to run validation was not accounting for `val_check_batch` being `inf` which occurs for iterable style datasets which don't have len defined. https://github.com/PyTorchLightning/pytorch-lightning/blob/490cc57809ebeba19003b4101393a8a058217c31/pytorch_lightning/trainer/data_loading.py#L288-L290

I think the dataloading logic is too complicated here. At the very least we need to consolidate logic across the dataloading mixin, debug connector, and training loop. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -233,60 +233,135 @@ def test_dataloaders_passed_to_fn(tmpdir, ckpt_path, n):
...
-    model = EvalModelTemplate()
-    model.train_dataloader = model.train_dataloader__infinite
-    model.val_dataloader = model.val_dataloader__infinite
-    model.test_dataloader = model.test_dataloader__infinite
 
+    ckpt_callback = ModelCheckpoint(monitor="val_log", save_top_k=1, mode="max", verbose=False)
+    epoch_cb = Counter()
+    epochs = 2
     trainer = Trainer(
-        default_root_dir=tmpdir,
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:276]: default root dir got lost

File: pytorch_lightning/callbacks/progress.py [hunk 2]
```diff
@@ -402,12 +403,14 @@ def on_train_epoch_start(self, trainer, pl_module):
             val_checks_per_epoch = total_train_batches // trainer.val_check_batch
             total_val_batches = total_val_batches * val_checks_per_epoch
         total_batches = total_train_batches + total_val_batches
-        reset(self.main_progress_bar, total_batches)
+        reset(self.main_progress_bar, convert_inf(total_batches))
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress.py:406]: I already had convert_inf in reset, it wouldn't do anything new here.

File: pytorch_lightning/callbacks/progress.py [hunk 3]
```diff
@@ -418,11 +421,11 @@ def on_validation_start(self, trainer, pl_module):
         else:
             self._update_bar(self.main_progress_bar)  # fill up remaining
             self.val_progress_bar = self.init_validation_tqdm()
-            reset(self.val_progress_bar, self.total_val_batches)
+            reset(self.val_progress_bar, convert_inf(self.total_val_batches))
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress.py:424]: see other comment

File: pytorch_lightning/callbacks/progress.py [hunk 4]
```diff
@@ -496,8 +499,8 @@ def _update_bar(self, bar: Optional[tqdm]) -> None:
 
 
 def convert_inf(x: Optional[Union[int, float]]) -> Optional[Union[int, float]]:
-    """ The tqdm doesn't support inf values. We have to convert it to None. """
-    if x == float('inf'):
+    """ The tqdm doesn't support inf/nan values. We have to convert it to None. """
+    if x is None or math.isinf(x) or math.isnan(x):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress.py:503]: it's probably nan here because some arithmetic with inf produces it. 
did you observe where that happens?

File: tests/trainer/test_dataloaders.py [hunk 5]
```diff
@@ -233,60 +233,135 @@ def test_dataloaders_passed_to_fn(tmpdir, ckpt_path, n):
...
+            self.test_epoch_count += 1
+
+
 @pytest.mark.parametrize(['limit_train_batches', 'limit_val_batches', 'limit_test_batches'], [
-    (0.0, 0.0, 0.0),
+    (0., 0., 0.),
     (1.0, 1.0, 1.0),
 ])
 def test_inf_dataloaders_with_limit_percent_batches(tmpdir, limit_train_batches, limit_val_batches, limit_test_batches):
-    """Verify inf train, val & test dataloaders (e.g. IterableDataset) passed with batch limit in percent"""
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:241]: can we keep/update the test description?

File: pytorch_lightning/trainer/training_loop.py [hunk 6]
```diff
@@ -825,19 +825,36 @@ def should_accumulate(self):
...
+        if (self.trainer.current_epoch + 1) % self.trainer.check_val_every_n_epoch != 0:
+            return False
 
-        should_check_val = ((is_val_check_batch and epoch_end_val_check) or self.trainer.should_stop
-                            or is_last_batch_for_infinite_dataset
-                            ) if on_epoch else (is_val_check_batch and not epoch_end_val_check)
+        # val_check_batch is inf for iterable datasets with no length defined
+        # TODO: the training loop should maintain this logic
+        # around limit_train_batches and val_check_batch
+        # not the trainer dataloading mixin
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:841]:
```suggestion
        # TODO: let training/eval loop handle logic around limit_*_batches and val_check_batch
```

File: tests/trainer/test_dataloaders.py [hunk 7]
```diff
@@ -233,60 +233,135 @@ def test_dataloaders_passed_to_fn(tmpdir, ckpt_path, n):
...
+        self.train_batches_seen = 0
+        self.val_batches_seen = 0
+        self.test_batches_seen = 0
+
+    def on_train_batch_start(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
+        if not trainer.sanity_checking:
+            self.train_batches_seen += 1
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        if not trainer.sanity_checking:
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:262]: let's turn sanity check off in the test and remove this if? simpler logic and the counter counts all seen. in any case, sanity check would not call this hook

**Review by awaelchli — APPROVED** [review]
lgtm, 
just minor comments for improvement

File: pytorch_lightning/callbacks/progress.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/callbacks/progress.py:503]: my guess is `self.total_train_batches + self.total_val_batches` in `on_train_batch_end` since we were missing the float(inf) check there? i'll see if i can pinpoint this better

Comment (awaelchli) [reply pytorch_lightning/callbacks/progress.py:503]: no it can't be because these numbers are all positive. it must be one of these cases where the arithmetic is undefined:
inf - inf
inf / inf
x / 0

Comment (ananthsub) [reply pytorch_lightning/callbacks/progress.py:503]: @awaelchli 

following the repro from #7303 

```
    ckpt_callback = ModelCheckpoint(monitor="val_log", save_top_k=1, mode="max", verbose=False)
    epoch_cb = Counter()
    epochs = 2
    trainer = Trainer(
        default_root_dir=tmpdir,
        num_sanity_val_steps=0,
        max_epochs=epochs,
        callbacks=[epoch_cb, ckpt_callback],
        limit_train_batches=3,
    )
    model = DummyModel()
    batch_size = 8
    train_dl = DataLoader(dataset=RandomIterableDataset(32, 128), batch_size=batch_size)
    val_dl = DataLoader(dataset=RandomIterableDataset(32, 128), batch_size=batch_size)
    trainer.fit(model, train_dataloader=train_dl, val_dataloaders=val_dl)
```

in the progress callback:
- total_train_batches=3
- total_val_batches=inf
- we only check if total_train_batches != inf, not val_batches
- trainer.val_check_batch is inf
- val_checks_per_epoch = 3 / inf -> 0.0
- total_val_batches = 0.0 * inf -> nan

Comment (awaelchli) [reply pytorch_lightning/callbacks/progress.py:503]: true, 0*inf is another case :))

do you think we should handle the logic here rather than in the progress bar code?

Comment (ananthsub) [reply pytorch_lightning/callbacks/progress.py:504]: We can do both, just in case non-number inputs are somehow passed to tqdm in another path. I also updated the check inside the progress callback in case total_val_batches is also float inf

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -825,19 +825,34 @@ def should_accumulate(self):
...
+        if (self.trainer.current_epoch + 1) % self.trainer.check_val_every_n_epoch != 0:
+            return False
 
-        should_check_val = ((is_val_check_batch and epoch_end_val_check) or self.trainer.should_stop
-                            or is_last_batch_for_infinite_dataset
-                            ) if on_epoch else (is_val_check_batch and not epoch_end_val_check)
+        # val_check_batch is inf for iterable datasets with no length defined
+        # TODO: let training/eval loop handle logic around limit_*_batches and val_check_batch
+        is_val_check_batch = False
+        if isinstance(self.trainer.limit_train_batches, int) and self.trainer.val_check_batch == float('inf'):
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/training_loop.py:841]: Love this refactor!

**Comment by kaushikb11** [comment]
@ananthsub There's a failing test other than the unrelated deepspeed failing test.

**Comment by ananthsub** [comment]
@kaushikb11 for some reason I cannot reproduce the test failure locally. 

but it's pointing out an issue from https://github.com/PyTorchLightning/pytorch-lightning/pull/6671 which is that the train loop force calling checkpointing on train end is bad because we cannot guarantee the monitor value is present

@awaelchli @carmocca any suggestions on how to debug in this case, or why the failure isn't consistent across versions?

File: pytorch_lightning/accelerators/accelerator.py [hunk 9]
```diff
@@ -114,7 +114,7 @@ def pre_dispatch(self, trainer: 'pl.Trainer') -> None:
     def _move_optimizer_state(self) -> None:
         """ Moves the state of the optimizers to the GPU if needed. """
         for opt in self.optimizers:
-            state = defaultdict(dict)
+            state: DefaultDict = defaultdict(dict)
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:117]: sent this out as a different PR here: #7318
but wanted it here to ensure tests were passing

**Review by ethanwharris — APPROVED** [review]
Awesome! LGTM :smiley:

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 8 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:841]: @kaushikb11 thanks! it still feels complicated to me. part of that is from `limit_train_batches` / `val_check_interval` having different types and possible meanings depending on both depending on the user input and dataloader specified. 

i'm wondering what's a better way to split "when to stop training mid-epoch" vs when to run validation or if a split is needed at all.

**Review by tchaton — APPROVED** [review]
LGTM !
